root/static/scripts/homepage/search.js

```diff
@@ -0,0 +1,218 @@
+/*
+ * @flow strict
+ * Copyright (C) 2025 MetaBrainz Foundation
+ *
+ * This file is part of MusicBrainz, the open internet music database,
+ * and is licensed under the GPL version 2, or (at your option) any
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import {faChevronDown, faXmark} from '@fortawesome/free-solid-svg-icons';
+import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
+import * as React from 'react';
+
+import dataProviderImage from '../../images/homepage/data-provider.png';
+import ethicalSourceImage from '../../images/homepage/ethical-forever.png';
+import openSourceImage from '../../images/homepage/open-source.png';
+import searchIcon from '../../images/homepage/search-bar-icon.svg';
+import {l} from '../common/i18n.js';
+
+import Blob from './blob.js';
+import {type WeeklyStatsT} from './stats.js';
+import entities from './utils.js';
+
+component Search (
+  weeklyStats: $ReadOnlyArray<WeeklyStatsT>,
+) {
+  const [searchQuery, setSearchQuery] = React.useState('');
+
+  const entitiesWithStats = entities.map((entity) => {
+    let statKey = `count.${entity.value}`;
+    if (entity.value === 'release_group') {
+      statKey = 'count.releasegroup';
+    }
+
+    const stat = weeklyStats.find((s) => s.stat === statKey);
+    return {...entity, stat};
+  });
+
+  const [selectedEntity, setSelectedEntity] = React.useState(
+    entitiesWithStats[0],
+  );
+  const [isModalOpen, setIsModalOpen] = React.useState(false);
+
+  const placeholder = selectedEntity.stat && selectedEntity.stat.total > 0
+    ? `Search ${selectedEntity.stat.total.toLocaleString()} ${
+      selectedEntity.stat?.name || selectedEntity.name
+    }...`
+    : `Search ${selectedEntity.name.toLowerCase()}...`;
+
+  const handleSearch = (e: SyntheticEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    window.location.href = `/search?query=${encodeURIComponent(
+      searchQuery,
+    )}&type=${selectedEntity.value}`;
+  };
+
+  const scrollToElement = () => {
+    const openSourceElement = document.getElementById('open-source');
+    if (openSourceElement) {
+      openSourceElement.scrollIntoView({behavior: 'smooth'});
+    }
+  };
+
+  const handleEntitySelect = (entity: typeof entitiesWithStats[0]) => {
```

Comment:
The typing here feels a bit hacky - might be better to have a defined flow type, and then you can also make that the expected type for `entitiesWithStats` when defining it for extra clarity? But @mwiencek might disagree.

Reply:
I've defined this type.

Reply:
Separate type sounds good. Also just to note, since I changed this elsewhere: we generally name types with a `T` suffix (e.g. `ReleaseGroupT`) to distinguish them from actual code, since component and class names also use UpperCamelCase.